frontend/apps/tic-tac-toe/src/features/tic-tac-toe/hooks.ts

```diff
@@ -185,20 +187,32 @@ export function useSubscriptionOnGameMessage(meta: ProgramMetadata) {
   };
 }
 
-export const useHandleCalculateGas = (address: HexString, meta: ProgramMetadata) => {
-  const { availableBalance } = useAccountAvailableBalance();
-  const calculateGasNative = useCalculateGasNative(address, meta);
-
-  const alert = useAlert();
+export const useCheckGaslessVouher = <T extends unknown>(callback: (params: T) => void) => {
+  const { account } = useAccount();
+  const signless = useSignlessTransactions();
+  const gasless = useGaslessTransactions();
+  const { isBalanceReady } = useBalance(gasless.voucherId);
 
-  return (initPayload: AnyJson, value?: AnyNumber | undefined) => {
-    const balance = Number(withoutCommas(availableBalance?.value || ''));
-    const existentialDeposit = Number(withoutCommas(availableBalance?.existentialDeposit || ''));
+  const [isNeedExecuteCallback, setIsNeedExecuteCallback] = useState(false);
+  const [savedParams, setSavedParams] = useState<T>();
 
-    if (!balance || balance < existentialDeposit) {
-      alert.error(`Low balance when calculating gas`);
+  useEffect(() => {
+    if (isNeedExecuteCallback && gasless.isEnabled && gasless.voucherId && isBalanceReady) {
+      setIsNeedExecuteCallback(false);
+      callback(savedParams as T);
+      setSavedParams(undefined);
+    }
+  }, [gasless.isEnabled, gasless.voucherId, isBalanceReady, isNeedExecuteCallback, callback, savedParams]);
+
+  const checkGaslessVoucher = async (params: T) => {
+    if (account && gasless.isEnabled && !gasless.voucherId && !signless.isActive) {
+      await gasless.requestVoucher(account.address);
+      setSavedParams(params);
+      setIsNeedExecuteCallback(true);
+    } else {
+      callback(params);
```

Comment:
I don't quite follow the logic here, why something simple like that wouldn't be enough?

```suggestion
  const checkGaslessVoucher = async (params: T) => {
    if (account && gasless.isEnabled && !gasless.voucherId && !signless.isActive) {
      await gasless.requestVoucher(account.address);
    }
    callback(params);
```

Reply:
This won’t work because checkBalance in the closure remembers the user's balance value. The voucher balance value is not yet ready immediately after requesting the voucher. So, here what we mainly wait for is isBalanceReady.

Reply:
We can resolve the situation differently. Two options:

1) Asynchronously check the balance in checkBalance if isBalanceReady === false.
2) Request the voucher upon pressing 'Enable gasless' instead of 'Start game'.